### DISCUSSION TITLE
How to create an autoincrement column when using UUID?

### DISCUSSION THREAD

**Discussion by Jedliu** [mikro-orm/mikro-orm#3187] [Q&A] [answered]
Hi, 

I'm wondering how to create an autoincrement column when using UUID as the primary key. 

Thank you!

**Comment by B4nan** [comment]
You cant, with UUID you dont increment, you always generate new one, they should be _universally unique_, thats where the two _U_ letters are coming from in the _UUID_.

**Reply by Jedliu** [reply]
I mean two seperated columns, one is UUID and another one is autoincremented number. Is this possible?

**Reply by B4nan** [reply]
Not as a composite PK, that's not supported. TBH I don't see any point of doing that, both of them are guaranteed to be unique on their own.

**Reply by B4nan** [reply]
What driver are you using? With postgres you can use sequences/serial for this, that can be a regular property too.

**Reply by Jedliu** [reply]
Yes, I'm using PostgreSQL. Would you please give me a clue on how to use sequences/serial? I didn't find it in the doc.

**Reply by B4nan** [reply]
i think this should be actually enough in postgres:

```ts
@Property({ autoincrement: true })
myProp!: number;
```

**Reply by Jedliu** [reply]
Just did the test, and it should be like this:
```typescript
@Property({ autoincrement: true, primary: false })
_id!: number;
```
Thank you so much for your help!

**Comment by Jedliu** [comment]
For anyone who is interested, this is the solution in Postgres with @B4nan's help:
```typescript
@Property({ autoincrement: true, primary: false })
_id!: number;
```

**Comment by Jedliu** [comment]
There are some problems during the migration when used in NestJS.
I created a repo [nestjs-mikroorm-demo](https://github.com/Jedliu/nestjs-mikroorm-demo-1/tree/20220608-autoincrement). Do the following steps to reproduce the problem.
```
npm run start:debug
```
Everything is working.
Then uncomment the _id property definition in `book.entity.ts`
There will be an error.
```
DriverException: alter table "book" add column "_id" serial primary key; - multiple primary keys for table "book" are not allowed
```
I have to drop all the tables and then the migration works.
However the database scheme is not correct. The _id becomes part of the primary key.

![book1](https://user-images.githubusercontent.com/771703/172510114-f30e3551-f386-4809-bd10-15e5e07976ca.jpg)

Save the `book.entity.ts` again to trigger the reload of the migration and then the database schema is correct.

![book2](https://user-images.githubusercontent.com/771703/172510395-11aac86d-01e5-49f1-b46c-0c362fe77ee2.jpg)

Same thing happens when update the definition of the second entity `comment.entity.ts`.
Could you please advise what the problem is?
Thank you,

**Reply by B4nan** [reply]
Weird, I remember we had tests for this but can't find any. It's indeed broken, will try to look into this during the day. 

What you can do now is:

```
@Property({ columnType: 'serial' })
_id!: number;
```

It will generate the table correctly, but it will keep generating additional migrations afterwards that would switch the column to a regular int without the sequence default. I even remember testing this case and adding a special handing for diffing serial columns 🤷

**Reply by Jedliu** [reply]
Update to the below codes and now it works. I haven't worked with the migrations yet so not noticed the problem you mentioned. Look forward to your fix. Thank you,
```typescript
@Property({ columnType: 'serial', autoincrement: true })
_id!: number;
```

**Reply by Jedliu** [reply]
Just now I found it still didn't work. The problem is there.
After dropping the tables, the first migration is not correct, and save the entity to trigger the migration again then it works.

**Reply by B4nan** [reply]
try latest dev version, should work with just `@Property({ autoincrement: true })`

https://github.com/mikro-orm/mikro-orm/commit/088afdb96ef8fd20a3868c050bb749a4e825fd19

**Reply by Jedliu** [reply]
Could you please advise how to use the latest dev version to test? Thank you!

**Comment by Jedliu** [comment]
Hi, I noticed the lasted release v5.2.0. Thank you for your dedicated work.

However, the problem still exists. The primary key definition for the **id** was lost. So I saw the error 
```
DriverException: alter table "comment" add constraint "comment_book_id_foreign" foreign key ("book_id") references "book" ("id") on update cascade; - there is no unique constraint matching given keys for referenced table "book"
```
I tried to create the migration files and found the alter to add the primary key part was lost.

Steps to reproduce the problem: [repo link](https://github.com/Jedliu/nestjs-mikroorm-demo-1/tree/20220608-autoincrement)
1- Drop all the tables
2- **npm run start:debug**

However, if I add the autoincrement definition without dropping all the tables, it works.

Thanks for your attention!

**Comment by Jedliu** [comment]
Finally, I got time to look into this again. I found where the problem is. 
In your test `tests/features/schema-generator/serial-property.postgres.test.ts`, please try this entity definition.

```ts
@Entity({ tableName: 'something' })
export class Something5 {

  @PrimaryKey({ primary: true })
  id!: string;

  @Property({ autoincrement: true })
  _id!: number;

  @Property()
  foo!: string;
}
```
The schema created is 
```
create table "something" ("id" varchar(255) not null, "_id" serial, "foo" varchar(255) not null);
```
The correct one should be
```
create table "something" ("id" varchar(255) not null, "_id" serial, "foo" varchar(255) not null, constraint "something_pkey" primary key ("id"));
```
The index was omitted to create in function
```
private createTable(tableDef: DatabaseTable): Knex.SchemaBuilder
```
Would you please look into this again to fix the logic of creating the index?

Thank you!

**Comment by Jedliu** [comment, marked as answer]
@B4nan You're amazing. Thanks for the quickest fix [PK constraint not created #3350](https://github.com/mikro-orm/mikro-orm/issues/3350)
